services/py-api/src/utils.py

```diff
@@ -43,3 +48,42 @@ async def stop(self) -> None:
             raise RuntimeError("The AsyncClient has not been initialized!")
 
         await self._async_client.aclose()
+
+
+class JwtUtility:
+    """A class providing generic methods for encoding data in with a predefined format into a JWT token, or decoding a JWT
+    token into a predefined format (TypedDict)"""
+
+    @staticmethod
+    def encode_data[T: BaseTypedDict](data: T) -> str:
+        return str(encode(payload=data, key=environ["SECRET_KEY"], algorithm="HS256"))
+
+    @staticmethod
+    def decode_data[T: BaseTypedDict](token: str, schema: Callable[..., T]) -> Result[T, str]:
+        try:
+            decoded_token: dict[str, Any] = decode(jwt=token, key=environ["SECRET_KEY"], algorithms=["HS256"])
+
+            # schema.__annotations___.keys() gets all the defined fields of the TypedDict without initializing it
+            # this will help us see if all the fields that we have defined in the schema are present in the decoded
+            # jwt token
+            for key in schema.__annotations__.keys():
+                if key not in decoded_token:
+                    return Err("The decoded token does not correspond with the provided schema.")
+
+            # We check both sides since we dont know which of the dictionaries is bigger than the other
+            for key in decoded_token:
+                if key not in schema.__annotations__.keys():
+                    return Err("The decoded token does not correspond with the provided schema.")
```

Comment:
are we keeping the str errors, or should we return custom ones, as @asynchroza mentioned?

Reply:
I just talked to @asynchroza and he thinks that it is better to keep them like this .